draft-ietf-tls-esni.md

```diff
@@ -446,33 +491,120 @@ to harmonize these to make sure that we maintain key separation.]]
+Section 9.2.)
+
+### Handling the server response {#handle-server-response}
+
+If the server negotiates TLS 1.3 or above and provides an
+"encrypted_server_name" extension in EncryptedExtensions, the client
+then processes the extension's "response_type" field:
+
+- If the value is "esni_accept", the client MUST check that the extension's
+  "nonce" field matches ClientESNIInner.nonce and otherwise abort the
+  connection with an "illegal_parameter" alert. The client then proceeds
+  with the connection as usual, verifying the certificate against the desired
+  name.
+
+- If the value is "esni_retry_request", the client proceeds with the handshake,
+  verifying the certificate against ESNIKeys.public_name as described in
+  {{verify-public-name}}. If verification or the handshake fails, the client
+  MUST return a failure to calling application. It MUST NOT use the retry keys
+  as described below.
+
+  Otherwise, when the handshake completes successfully with the public name
+  verified, the client MUST abort the connection with an "esni_required" alert.
+  The client can then regard the ESNI keys as securely replaced by the server.
+  It processes the "retry_keys" field from the server's "encrypted_server_name"
+  extension. If one of the values used a version known to the client, the client
+  SHOULD retry the handshake with a new transport connection, using that value
+  to encrypt the SNI. If no value is applicable, the client SHOULD retry with
+  ESNI disabled.
+
+  These retry keys may only be applied to the retry connection. The client MUST
+  continue to use the previously cached keys for subsequent connections. This
+  avoids introducing pinning concerns or a tracking vector, should a malicious
+  server present client-specific retry keys to identify clients.
+
+- If the field contains any other value, the client MUST abort the connection
+  with an "illegal_parameter" alert.
+
+If the server negotiates an earlier version of TLS, or if it does not
+provide an "encrypted_server_name" extension in EncryptedExtensions, the
+client proceeds with the handshake, verifying the certificate against
+ESNIKeys.public_name as described in {{verify-public-name}}. The client MUST
+NOT enable the False Start optimization {{RFC7918}} for this handshake. If
+verification or the handshake fails, the client MUST return a failure to the
+calling application. It MUST NOT treat this as a signal to disable ESNI as
+described below.
+
+Otherwise, when the handshake completes successfully with the public name
+verified, the client MUST abort the connection with an "esni_required" alert.
+The client can then regard ESNI as securely disabled by the server. It SHOULD
+retry the handshake with a new transport connection and ESNI disabled.
+
+[[TODO: Key replacement is significantly less scary than saying that ESNI-naive
+  servers bounce ESNI off. Is it worth defining a strict mode toggle in the ESNI
+  keys, for a deployment to indicate it is ready for that? ]]
```

Comment:
I'm not sure I understand this. Where would the toggle go, in DNS? And having the toggle turned on would signal to the client that it must not retry the connection with ESNI disabled?

Reply:
Some extension on the ESNIKeys yeah. This is a TODO, not spec text.